// posthog-js/lib/src/logs-defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveLogsConfig = resolveLogsConfig;
// Browser defaults. Faster flush than mobile since a tab can be suspended anytime.
var DEFAULT_FLUSH_INTERVAL_MS = 3000;
// Flush trigger — drain proactively once this many records are buffered.
var DEFAULT_MAX_BUFFER_SIZE = 100;
var DEFAULT_MAX_LOGS_PER_INTERVAL = 1000;
// Bounds each POST independently of the buffer so a full buffer drains as ~10 requests.
var DEFAULT_MAX_BATCH_RECORDS_PER_POST = 100;
/**
 * Translates the flat public `logs` config into the resolved shape the core
 * `PostHogLogs` consumes, applying browser defaults.
 */
function resolveLogsConfig(config) {
    var _a, _b, _c, _d, _e, _f;
    var flushIntervalMs = (_a = config === null || config === void 0 ? void 0 : config.flushIntervalMs) !== null && _a !== void 0 ? _a : DEFAULT_FLUSH_INTERVAL_MS;
    var maxBufferSize = (_b = config === null || config === void 0 ? void 0 : config.maxBufferSize) !== null && _b !== void 0 ? _b : DEFAULT_MAX_BUFFER_SIZE;
    var maxLogsPerInterval = (_c = config === null || config === void 0 ? void 0 : config.maxLogsPerInterval) !== null && _c !== void 0 ? _c : DEFAULT_MAX_LOGS_PER_INTERVAL;
    // Eviction cap at the rate cap (or buffer, if larger): hold everything the cap
    // admits before dropping oldest; flushing still triggers at `maxBufferSize`.
    var maxQueueSize = Math.max(maxBufferSize, maxLogsPerInterval);
    // Hoist OTLP keys from `resourceAttributes` into the named fields so they keep
    // their precedence through the core merge. No `service.name` default: core fills
    // `unknown_service` when unset, preserving the programmatic path's prior default.
    var resourceAttributes = config === null || config === void 0 ? void 0 : config.resourceAttributes;
    return {
        serviceName: (_d = resourceAttributes === null || resourceAttributes === void 0 ? void 0 : resourceAttributes['service.name']) !== null && _d !== void 0 ? _d : config === null || config === void 0 ? void 0 : config.serviceName,
        serviceVersion: (_e = resourceAttributes === null || resourceAttributes === void 0 ? void 0 : resourceAttributes['service.version']) !== null && _e !== void 0 ? _e : config === null || config === void 0 ? void 0 : config.serviceVersion,
        environment: (_f = resourceAttributes === null || resourceAttributes === void 0 ? void 0 : resourceAttributes['deployment.environment']) !== null && _f !== void 0 ? _f : config === null || config === void 0 ? void 0 : config.environment,
        resourceAttributes: resourceAttributes,
        beforeSend: config === null || config === void 0 ? void 0 : config.beforeSend,
        flushIntervalMs: flushIntervalMs,
        maxBufferSize: maxBufferSize,
        maxQueueSize: maxQueueSize,
        maxBatchRecordsPerPost: DEFAULT_MAX_BATCH_RECORDS_PER_POST,
        // Rate-cap window tracks the flush interval: the browser exposes a single
        // cadence knob and has no separate window setting.
        rateCapWindowMs: flushIntervalMs,
        maxLogsPerInterval: maxLogsPerInterval,
        // Mobile-only budgets; the browser drains on pagehide via sendBeacon.
        backgroundFlushBudgetMs: 0,
        terminationFlushBudgetMs: 0,
    };
}
