// posthog-js/lib/src/posthog-logs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostHogLogs = void 0;
var constants_1 = require("./constants");
var config_1 = __importDefault(require("./config"));
var core_1 = require("@posthog/core");
var globals_1 = require("./utils/globals");
var utils_1 = require("./utils");
var logger_1 = require("./utils/logger");
var logs_defaults_1 = require("./logs-defaults");
var LOGS_ENDPOINT = '/i/v1/logs';
// Safety backstop for a `_send_request` that never calls back. Set above the
// request layer's own 60s timeout so a real (slow-but-completing) request always
// settles via its callback first; this only fires on a genuinely callback-less
// send (e.g. request enqueued before load, or a transport that never reports).
var LOGS_SEND_TIMEOUT_MS = 90000;
var PostHogLogs = /** @class */ (function () {
    function PostHogLogs(_instance) {
        var _this = this;
        var _a;
        this._instance = _instance;
        this._isLogsEnabled = false;
        this._isLoaded = false;
        this._logger = (0, logger_1.createLogger)('[logs]');
        // In-memory only; records do not survive a page reload.
        this._queue = [];
        this._onReconnect = function () {
            var _a;
            (_a = _this._core) === null || _a === void 0 ? void 0 : _a.onReconnect();
        };
        if (this._instance && ((_a = this._instance.config.logs) === null || _a === void 0 ? void 0 : _a.captureConsoleLogs)) {
            this._isLogsEnabled = true;
        }
        // Flush promptly when the tab regains connectivity instead of waiting out
        // the retry backoff. One listener for the wrapper's lifetime, routed to the
        // current core (which may be rebuilt on a config change).
        if (globals_1.window) {
            (0, utils_1.addEventListener)(globals_1.window, 'online', this._onReconnect);
        }
    }
    // The extension is constructed before `init` applies config, so build the core
    // lazily and rebuild when `config.logs` is swapped (e.g. via `set_config`).
    // Reset the old core first so its armed timer can't double-flush the shared,
    // wrapper-owned queue. A flush already in flight on the old core can still
    // finish against that queue, so a config swap mid-flush may re-send a head
    // batch — a duplicate, never a loss.
    PostHogLogs.prototype._getCore = function () {
        var _this = this;
        var _a, _b, _c;
        var logsConfig = (_b = (_a = this._instance) === null || _a === void 0 ? void 0 : _a.config) === null || _b === void 0 ? void 0 : _b.logs;
        if (!this._core || this._resolvedFrom !== logsConfig) {
            (_c = this._core) === null || _c === void 0 ? void 0 : _c.reset();
            this._resolvedFrom = logsConfig;
            this._resolvedConfig = (0, logs_defaults_1.resolveLogsConfig)(logsConfig);
            this._core = new core_1.PostHogLogs(this._createHost(), this._resolvedConfig, this._logger, function () { return _this._getSdkContext(); }, function (fn) { return fn(); });
        }
        return this._core;
    };
    // `_getCore` assigns `_resolvedConfig` alongside `_core`, so it's set on return.
    PostHogLogs.prototype._getResolvedConfig = function () {
        this._getCore();
        return this._resolvedConfig;
    };
    PostHogLogs.prototype.initialize = function () {
        this.loadIfEnabled();
    };
    PostHogLogs.prototype.onRemoteConfig = function (response) {
        var _a;
        var logCapture = (_a = response.logs) === null || _a === void 0 ? void 0 : _a.captureConsoleLogs;
        if ((0, core_1.isNullish)(logCapture) || !logCapture) {
            return;
        }
        this._isLogsEnabled = true;
        this.loadIfEnabled();
    };
    PostHogLogs.prototype.reset = function () {
        var _a;
        this._queue = [];
        (_a = this._core) === null || _a === void 0 ? void 0 : _a.reset();
    };
    PostHogLogs.prototype.captureLog = function (options) {
        this._getCore().captureLog(options);
    };
    Object.defineProperty(PostHogLogs.prototype, "logger", {
        get: function () {
            var _this = this;
            if (!this._capture_logger) {
                this._capture_logger = {
                    trace: function (body, attributes) { return _this.captureLog({ body: body, level: 'trace', attributes: attributes }); },
                    debug: function (body, attributes) { return _this.captureLog({ body: body, level: 'debug', attributes: attributes }); },
                    info: function (body, attributes) { return _this.captureLog({ body: body, level: 'info', attributes: attributes }); },
                    warn: function (body, attributes) { return _this.captureLog({ body: body, level: 'warn', attributes: attributes }); },
                    error: function (body, attributes) { return _this.captureLog({ body: body, level: 'error', attributes: attributes }); },
                    fatal: function (body, attributes) { return _this.captureLog({ body: body, level: 'fatal', attributes: attributes }); },
                };
            }
            return this._capture_logger;
        },
        enumerable: false,
        configurable: true
    });
    // An explicit transport drains the whole queue in one request over that transport
    // (core's batched flush can't force a transport, and the unload sendBeacon must be
    // synchronous). No transport → core's batched, 413-aware, retrying flush.
    PostHogLogs.prototype.flushLogs = function (transport) {
        var _this = this;
        if (transport) {
            this._flushViaTransport(transport);
            return;
        }
        void this._getCore()
            .flush()
            .catch(function (err) { return _this._logger.error('PostHog logs flush failed:', err); });
    };
    PostHogLogs.prototype.loadIfEnabled = function () {
        var _this = this;
        if (!this._isLogsEnabled || this._isLoaded) {
            return;
        }
        var phExtensions = globals_1.assignableWindow === null || globals_1.assignableWindow === void 0 ? void 0 : globals_1.assignableWindow.__PosthogExtensions__;
        if (!phExtensions) {
            this._logger.error('PostHog Extensions not found.');
            return;
        }
        var loadExternalDependency = phExtensions.loadExternalDependency;
        if (!loadExternalDependency) {
            this._logger.error(constants_1.LOAD_EXT_NOT_FOUND);
            return;
        }
        loadExternalDependency(this._instance, 'logs', function (err) {
            var _a;
            if (err || !((_a = phExtensions.logs) === null || _a === void 0 ? void 0 : _a.initializeLogs)) {
                _this._logger.error('Could not load logs script', err);
            }
            else {
                phExtensions.logs.initializeLogs(_this._instance);
                _this._isLoaded = true;
            }
        });
    };
    // Host adapter for core's `PostHogLogs`; structurally checked against `LogsHost`
    // at the `new CorePostHogLogs` call, so no explicit annotation is needed.
    PostHogLogs.prototype._createHost = function () {
        var _this = this;
        var ph = this._instance;
        return {
            // The browser gates capture through `is_capturing()` (see `optedOut`).
            get isDisabled() {
                return false;
            },
            get optedOut() {
                return !ph.is_capturing();
            },
            // Live queue by reference; core mutates it in place and persists via the setter.
            getPersistedProperty: function (key) {
                return key === core_1.PostHogPersistedProperty.LogsQueue ? _this._queue : undefined;
            },
            setPersistedProperty: function (key, value) {
                var _a;
                if (key === core_1.PostHogPersistedProperty.LogsQueue) {
                    _this._queue = (_a = value) !== null && _a !== void 0 ? _a : [];
                }
            },
            _sendLogsBatch: function (payload) { return _this._sendLogsBatch(payload); },
            getLibraryId: function () { return config_1.default.LIB_NAME; },
            getLibraryVersion: function () { return config_1.default.LIB_VERSION; },
        };
    };
    PostHogLogs.prototype._sendLogsBatch = function (payload) {
        var _this = this;
        // eslint-disable-next-line compat/compat
        return new Promise(function (resolve) {
            var settled = false;
            var settle = function (outcome) {
                if (settled) {
                    return;
                }
                settled = true;
                clearTimeout(timer);
                resolve(outcome);
            };
            // Backstop for `_send_request` paths that never call back, so the promise
            // always settles and core's flush can't wedge. Keeps records for retry.
            var timer = setTimeout(function () { return settle({ kind: 'retry-later', error: new Error('logs request timed out') }); }, LOGS_SEND_TIMEOUT_MS);
            _this._instance._send_request({
                method: 'POST',
                url: _this._logsUrl(),
                data: payload,
                compression: 'best-available',
                batchKey: 'logs',
                // Notify on the drop paths (not loaded, rate limited) so they retry, not stall.
                fireCallbackOnDrop: true,
                callback: function (response) {
                    var _a;
                    var status = response.statusCode;
                    if (status >= 200 && status < 300) {
                        settle({ kind: 'ok' });
                    }
                    else if (status === 413) {
                        settle({ kind: 'too-large' });
                    }
                    else if (status === 0 || status === 429 || status >= 500) {
                        // Transient (network / rate-limit / server error): keep and retry.
                        settle({
                            kind: 'retry-later',
                            error: (_a = response.error) !== null && _a !== void 0 ? _a : new Error("logs request failed with status ".concat(status)),
                        });
                    }
                    else {
                        // Client error (4xx): won't succeed on retry, drop.
                        settle({ kind: 'fatal', error: new Error("logs request failed with status ".concat(status)) });
                    }
                },
            });
        });
    };
    PostHogLogs.prototype._flushViaTransport = function (transport) {
        var config = this._getResolvedConfig();
        if (this._queue.length === 0) {
            return;
        }
        var records = this._queue.map(function (e) { return e.record; });
        this._queue = [];
        // Shared with the core flush path so resource attributes can't drift.
        var payload = (0, core_1.buildOtlpLogsPayload)(records, (0, core_1.buildResourceAttributes)(config, config_1.default.LIB_NAME, config_1.default.LIB_VERSION), config_1.default.LIB_NAME, config_1.default.LIB_VERSION);
        this._instance._send_request({
            method: 'POST',
            url: this._logsUrl(),
            data: payload,
            compression: 'best-available',
            batchKey: 'logs',
            transport: transport,
        });
    };
    PostHogLogs.prototype._logsUrl = function () {
        return (this._instance.requestRouter.endpointFor('api', LOGS_ENDPOINT) +
            '?token=' +
            encodeURIComponent(this._instance.config.token));
    };
    PostHogLogs.prototype._getSdkContext = function () {
        var _a;
        var context = {};
        context.distinctId = this._instance.get_distinct_id();
        if (this._instance.sessionManager) {
            var sessionId = this._instance.sessionManager.checkAndGetSessionAndWindowId(true).sessionId;
            context.sessionId = sessionId;
        }
        if ((_a = globals_1.assignableWindow === null || globals_1.assignableWindow === void 0 ? void 0 : globals_1.assignableWindow.location) === null || _a === void 0 ? void 0 : _a.href) {
            context.currentUrl = globals_1.assignableWindow.location.href;
        }
        if (this._instance.featureFlags) {
            var flags = this._instance.featureFlags.getFlags();
            if (flags && flags.length > 0) {
                context.activeFeatureFlags = flags;
            }
        }
        return context;
    };
    return PostHogLogs;
}());
exports.PostHogLogs = PostHogLogs;
